source/atlas-data-lake-testing/tests/README.rst

```diff
@@ -0,0 +1,51 @@
+=====================
+Atlas Data Lake Tests
+=====================
+
+.. contents::
+
+----
+
+Introduction
+============
+
+The YAML and JSON files in this directory tree are platform-independent tests
+that drivers can use to assert compatibility with `Atlas Data Lake <https://docs.mongodb.com/datalake>`_.
+
+Running these integration tests will require a running ``mongohoused``
+with data available in its ``test.driverdata`` collection. See the
+`ADL directory in drivers-evergreen-tools <https://github.com/mongodb-labs/drivers-evergreen-tools/tree/master/.evergreen/atlas_data_lake>`_
+and `10gen/mongohouse README <https://github.com/10gen/mongohouse/blob/master/README.md>`_
+for more information.
+
+Several prose tests, which are not easily expressed in YAML, are also presented
+in this file. Those tests will need to be manually implemented by each driver.
+
+Test Format
+===========
+
+The same as the `CRUD Spec Test format <../../crud/tests/README.rst#Test-Format>`_.
+
+Prose Tests
+===========
+
+The following tests MUST be implemented to fully test compatibility with
+Atlas Data Lake.
+
+#. Test that the ``killCursors`` command works with Atlas Data Lake.
```

Comment:
@jstewart-mongo I thought that this turned out to be difficult to implement, because the test won't have any way of knowing what namespace to specify when constructing a killCursors command manually.

Reply:
For the Java driver, the namespace needed to pass to `killCursors` is in the batch cursor, which is not accessible from `MongoCursor`, so the Java driver has to revert to closing the cursor and checking the command listener to check if `killCursors` was executed. For other drivers, if access to the batch cursor is available, then the driver can call `killCursors` directly with the included namespace from the batch cursor.

Reply:
@jmikola do you expect any drivers to be able to access the namespace like that?  I would think that for most it would be hidden inside a cursor object that is not accessible from tests.

Reply:
I glossed over "batch cursor" in John's original comments. Is that something specific to the Java driver?

I assumed on my earlier review that this would be like `getMore` commands, which always used the "cursors" database; however, that may not be the case and I remember now that [`killCursors`](https://docs.mongodb.com/manual/reference/command/killCursors/) also requires a collection name.

Is there a concrete example of what needs to happen here? For instance, if the driver executes a `find` on `test.driverdata`, what do we expect the returned response to `find` and the subsequent `killCursors` command to look like? I assume "batch cursor" is referring to details we'd observe in the response to `find`, so perhaps drivers could rely on APM to capture that if their cursor doesn't provide a public API to get the namespace we need for `killCursors`.

Reply:
Let's add a new paragraph under this list item:

> Drivers that need to manually execute a `killCursors` command for this test should note that the namespace of the cursor returned by ADL (and expected for `killCursors` and `getMore` commands) will differ from the namespace of the original `find` operation. Drivers may need to utilize APM to capture the `find` command response, access this namespace from the `cursor.ns` field, and extract database and collection names to use for the `killCursors` command.

Reply:
Ah, @jyemin explained that this is less about testing ADL receives a `killCursors` command and more about the driver properly constructing any `killCursors` command it might send. I'm not aware of any drivers that have public APIs to send `killCursors`, so let's focus on the general case of a driver that sends it when a cursor object is destroyed and the driver cannot definitively conclude that the server has already closed it (e.g. already in the final batch, or iterated it to completion).

Given that, I think this prose test should do the following:

 * Execute `find` with the necessary options to guarantee that full iteration would require a `getMore`
 * Drivers can use APM to capture the `find` response and extract the namespace for the internal cursor (aka "batch cursor")
 * The test then destroys the cursor object and uses APM to assert that the driver sends a `killCursors` command with the appropriate namespace. They can also assert the response of this command to verify that the cursor was actually killed in ADL.
 * The driver can manually construct a `getMore` command and execute it with `runCommand` and expect an error, since the cursor will not exist. Manually executing `getMore` is likely necessary because the iterable cursor object will have been destroyed to accomplish the previous step of triggering `killCursors`.

This can be repeated/adapted for any code paths for which the driver might issue a `killCursors` command. I expect for most drivers, there's only one code path when destroying a cursor, but we can leave the wording open to refer to hypothetical public APIs for killing the cursor.